### PR TITLE
x86 optimization for gemm unified elempack

### PR DESCRIPTION
**Pull request by nihui**
- [x] optimize C offset calculation
- [ ] optimize load C
- [x] optimize 2x12/2x8/2x4
- [x] optimize 1x12/1x8/1x4
- [x] optimize pack A
- [x] optimize transpose A
- [x] optimize pack B
- [x] optimize transpose B
- [x] add openmp threads
- [x] replace unaligned load/store with aligned version
- [x] gemm with constant A pre-packed/transposed
- [x] gemm with constant B pre-packed/transposed
- [ ] new cpu api for getting cache size
- [ ] cache size / threads aware TILE M/N/K size
- [x] size-channel tensor as input
- [x] size-channel tensor as output
- [ ] reduce memory usage
- [x] test coverage++

### CODE REVIEW

File: src/layer/x86/gemm_x86.cpp [hunk 1]
```diff
@@ -0,0 +1,5487 @@
...
+
+    Mat tmp;
+    if (K > TILE_K)
+        tmp.create(TILE_N, TILE_M, 4u, opt.blob_allocator);
+
+    int i = 0;
+    for (; i < M; i += TILE_M)
+    {
+        int j = 0;
+        for (; j < N; j += TILE_N)
```
Comment (tpoisonooo) [inline src/layer/x86/gemm_x86.cpp:119]: 为啥我总感觉顺序应该是 TILE_M  -->  TILE_K --> TILE_N

Comment (tpoisonooo) [reply src/layer/x86/gemm_x86.cpp:119]: 对应代码在这里  https://github.com/tpoisonooo/how-to-optimize-gemm/blob/41352768008263f5027d6c2ef3b863710b377139/aarch64/MMult_4x4_18.cpp#L92

**Review by tpoisonooo — APPROVED** [review]
up 无敌，所向披靡！

File: src/layer/x86/gemm_x86.cpp [hunk 2]
```diff
@@ -0,0 +1,5487 @@
...
+                    {
+                        // transpose_pack_a(A, AT, TILE_M, TILE_K, opt);
+                        const int elempack = A.elempack;
+
+                        float* pp = AT.row(k / TILE_K);
+
+                        int ii = 0;
+#if __SSE2__
+#if __AVX__
+#if __AVX512F__
```
[the remark this answers is not in the document]
Comment (nihui) [reply src/layer/x86/gemm_x86.cpp]: resolved

**Pull request merged**